### PR TITLE
Ifpack2: Fuse all Chebyshev kernels into one

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/ifpack2 

## Motivation
Fuse ScaledDampedResidual and update kernels into one Chebyshev kernel. This is a blatant rip-off of @mhoemmen's work.

### CODE REVIEW

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+template<class ExecutionSpace>
+int64_t
+chebyshev_kernel_vector_launch_parameters (int64_t numRows,
+                                                 int64_t nnz,
+                                                 int64_t rows_per_thread,
+                                                 int& team_size,
+                                                 int& vector_length)
+{
+  using execution_space = typename ExecutionSpace::execution_space;
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp]: Later TODO: Figure out if these parameters are actually good.  Also, I would have preferred returning a struct with named fields, to the current "return a value and have two output parameters" approach.

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+  if (vector_length < 1) {
+    vector_length = 1;
+    while (vector_length<32 && vector_length*6 < nnz_per_row) {
+      vector_length *= 2;
+    }
+  }
+
+  // Determine rows per thread
+  if (rows_per_thread < 1) {
+#ifdef KOKKOS_ENABLE_CUDA
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp]: We have way too much `#ifdef KOKKOS_ENABLE_CUDA` in Trilinos -- not sure how to generalize that here but just something to keep in mind.

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+  const char kernel_label[] = "chebyshev_kernel_vector";
+  policy_type policy (1, 1);
+  if (team_size < 0) {
+    policy = policy_type (worksets, Kokkos::AUTO, vector_length);
+  }
+  else {
+    policy = policy_type (worksets, team_size, vector_length);
+  }
+
+  // Canonicalize template arguments to avoid redundant instantiations.
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp]: We should actually canonicalize them to their nonowning variants (`MemoryTraits<Unmanaged>`).

File: packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+template<class TpetraOperatorType>
+void
+ChebyshevKernel<TpetraOperatorType>::
+setMatrix (const Teuchos::RCP<const operator_type>& A)
+{
+  if (A_op_.get () != A.get ()) {
+    A_op_ = A;
+
+    // We'll (re)allocate these on demand.
+    X_colMap_ = std::unique_ptr<vector_type> (nullptr);
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp]: I wish there were a way to get at the column-Map MultiVector that already lives in CrsMatrix.

**Comment by cgcgcg** [comment]
@jhux2 Both in Ifpack2 and in MueLu. (Hence all the failing tests in the previous iterations of this change.)

File: packages/ifpack2/src/Ifpack2_Parameters.cpp [hunk 5]
```diff
@@ -64,6 +67,17 @@ void getValidParameters(Teuchos::ParameterList& params)
   params.set("fact: iluk level-of-fill", (int)1);
   params.set("fact: iluk level-of-overlap", (int)0);
 
+  // Ifpack2_Chebyshev
+  params.set("chebyshev: max eigenvalue",STS::nan());
+  params.set("chebyshev: ratio eigenvalue",STS::nan());
+  params.set("chebyshev: min eigenvalue",(double)30.0);
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Parameters.cpp]: 1. Avoid C-style casts; use C++ casts (either `double(30.0)` or `static_cast<double>(30.0)` -- I prefer the latter for integer casts, which may be a source of bugs).
2. The type of the literal `30.0` is `double`.

File: packages/ifpack2/src/Ifpack2_Parameters.cpp [hunk 6]
```diff
@@ -64,6 +67,17 @@ void getValidParameters(Teuchos::ParameterList& params)
...
+  params.set("chebyshev: degree",(int)1);
+  params.set("chebyshev: eigenvalue max iterations",(int)10);
+  params.set("chebyshev: assume matrix does not change",false);
+  // params.set("chebyshev: operator inv diagonal",Teuchos::null);
+  params.set("chebyshev: min diagonal value",STS::eps());
```
Comment (mhoemmen) [inline packages/ifpack2/src/Ifpack2_Parameters.cpp]: Is the "min diagonal value" a magnitude (as `eps()` is) or a value?  That is, if `Scalar` is complex, should the min diagonal value be complex?

File: packages/ifpack2/src/Ifpack2_Parameters.cpp (hunk 5 above)
Comment (cgcgcg) [reply packages/ifpack2/src/Ifpack2_Parameters.cpp]: Have a look at the rest of the file. Just trying to be consistent here. (But we should change a lot of that once we rework that, I agree.)

File: packages/ifpack2/src/Ifpack2_Parameters.cpp (hunk 6 above)
Comment (cgcgcg) [reply packages/ifpack2/src/Ifpack2_Parameters.cpp]: The file assumes `Scalar=double`, so that's ok here.

**Pull request merged**